File: mcu/arduino/programs/SerialPort.py
'''
This program runs two concurrent character pipes (one for input and one for 
output) to the specified serial device at a specified baudrate. The serial
device file path is given as the first command line argument while the baudrate
is given as the second command line argument.

Example usage:
$ python SerialPort.py /dev/cu.usbmodem1421 115200
'''

import sys
import serial
import threading

serial_port = None
closing = False

def run_outgoing_pipe():
    try:
        while True:
            data = sys.stdin.readline()
            data = data[:len(data)-1]
            if data == 'q':  # exit program
                break
            
            serial_port.write(data)
            serial_port.flush()
    except:
        pass

def run_incoming_pipe():
    try:
        while True:
            data = serial_port.read(size=1)
            if closing:
                break
            if len(data) < 1:  # read timeout
                continue
            
            sys.stdout.write(data + '\n')
            sys.stdout.flush()
    except:
        pass

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Exactly two command line arguments are required.")

    serial_port = serial.Serial(sys.argv[1], int(sys.argv[2]))
    serial_port.timeout = 0.1;

    outpipe_thread = threading.Thread(target=run_outgoing_pipe)
    inpipe_thread = threading.Thread(target=run_incoming_pipe)

    outpipe_thread.start()
    inpipe_thread.start()

    outpipe_thread.join()
    closing = True
    inpipe_thread.join()
    serial_port.close()

File: mcu/arduino/programs/SerialPort.cpp
/**
 * This program runs two concurrent character pipes (one for input and one for
 * output) to the specified serial device at a specified baudrate. The serial
 * device file path is given as the first command line argument while the baudrate
 * is given as the second command line argument.
 *
 * Example usage:
 * $ ./SerialPort.cpp /dev/cu.usbmodem1421 115200
 */

#include <stdlib.h>
#include <unistd.h>

#include <iostream>
#include <thread>
#include "serial.h"

using std::cerr;
using std::cout;
using std::endl;
using std::cin;
using std::string;

serial_t uart_port;
uint8_t send_buf[128];
uint8_t recv_buf[128];

bool closing = false;

void run_outgoing_pipe()
{
  uint8_t buf[128];
  string command;
  while (true)
  {
    cout << "waiting for command" << endl;
    cin >> command;
    if (command == "q")
      break;

    cout << "outgoing command" << endl;

    // send command string excluding the terminating null byte
    if (serial_write(&uart_port, (const uint8_t*)command.c_str(), sizeof(command.c_str()-1)) < 0) {
        cerr << "serial_write(): %s\n" << serial_errmsg(&uart_port) << endl;
        break;
    }
    serial_flush(&uart_port);
  }
}

void run_incoming_pipe()
{
  uint8_t buf[128];
  unsigned int *bytes_waiting;
  int bytes_read;
  while (true)
  {
    cout << "start poll" << endl;
    serial_poll(&uart_port, 1000);
    cout << "end poll" << endl;
    serial_input_waiting(&uart_port, bytes_waiting);
    if (*bytes_waiting > 0)
    {
      if ((bytes_read = serial_read(&uart_port, buf, *bytes_waiting, 50)) < 0) {
          cerr << "serial_read(): %s\n" << serial_errmsg(&uart_port) << endl;
          exit(1);
      }
      cout << "bytes read" << endl;
    }
  }
}

int main(int argc, char **argv)
{
  cout << "main - 1" << endl;
  if (argc != 3)
  {
    std::cerr << "Exactly two command line arguments are required." << endl;
    return 1;
  }

  cout << "main - 2" << endl;
  // open device with specified baudrate, and defaults of 8N1 with no flow control
  if (serial_open(&uart_port, argv[1], std::atoi(argv[2])) < 0) {
      cerr << "serial_open(): %s\n" << serial_errmsg(&uart_port) << endl;
      return 1;
  }

  cout << "main - 3" << endl;
  std::thread outpipe_thread(run_outgoing_pipe);
  std::thread inpipe_thread(run_incoming_pipe);

  cout << "main - 4" << endl;
  outpipe_thread.join();

  cout << "main - 5" << endl;
  closing = true;
  inpipe_thread.join();

  cout << "main - 6" << endl;
  serial_close(&uart_port);

  return 0;
}
